src/couch_stats/src/couch_stats_resource_tracker.erl

```diff
@@ -0,0 +1,1014 @@
+        btree_folds => ChangesProcessed,
+        changes_returned => ChangesReturned,
+        ioq_calls => IoqCalls
+    }.
+
+
+convert_mfa({M0, F0, A0}) ->
+    M = atom_to_binary(M0),
+    F = atom_to_binary(F0),
+    A = integer_to_binary(A0),
+    <<M/binary, ":", F/binary, "/", A/binary>>;
+convert_mfa(undefined) ->
+    null.
+
+convert_type(Atom) when is_atom(Atom) ->
+    atom_to_binary(Atom);
+convert_type({coordinator, Verb0, Path0}) ->
+    Verb = atom_to_binary(Verb0),
+    Path = list_to_binary(Path0),
+    <<"coordinator:", Verb/binary, ":", Path/binary>>;
+convert_type({worker, M0, F0}) ->
+    M = atom_to_binary(M0),
+    F = atom_to_binary(F0),
+    <<"worker:", M/binary, ":", F/binary>>.
+
+get_pid_ref() ->
+    case get(?PID_REF) of
+        undefined ->
+            Ref = make_ref(),
+            set_pid_ref({self(), Ref});
+        PidRef ->
+            PidRef
+    end.
+
+
+create_context() ->
+    is_enabled() andalso create_context(self()).
+
+
+create_context(Pid) ->
+    case is_enabled() of
+        false ->
+            ok;
+        true ->
+            Ref = make_ref(),
+            Rctx = make_record(Pid, Ref),
+            track(Rctx),
+            create_resource(Rctx),
+            Rctx
+    end.
+
+
+create_resource(#rctx{} = Rctx) ->
+    %% true = ets:insert(?MODULE, Rctx).
+    catch ets:insert(?MODULE, Rctx).
+
+%% add type to disnguish coordinator vs rpc_worker
+create_context(From, {M,F,_A} = MFA, Nonce) ->
+    case is_enabled() of
+        false ->
+            ok;
+        true ->
+            PidRef = get_pid_ref(), %% this will instantiate a new PidRef
+            %% TODO: extract user_ctx and db/shard from
+            Rctx = #rctx{
+                pid_ref = PidRef,
+                from = From,
+                mfa = MFA,
+                type = {worker, M, F},
+                nonce = Nonce
+            },
+            track(Rctx),
+            erlang:put(?DELTA_TZ, Rctx),
+            create_resource(Rctx),
+            Rctx
+    end.
+
+create_coordinator_context(#httpd{path_parts=Parts} = Req) ->
+    is_enabled() andalso create_coordinator_context(Req, io_lib:format("~p", [Parts])).
+
+create_coordinator_context(#httpd{} = Req, Path) ->
+    case is_enabled() of
+        false ->
+            ok;
+        true ->
+            #httpd{
+                method = Verb,
+                nonce = Nonce
+            } = Req,
+            PidRef = get_pid_ref(), %% this will instantiate a new PidRef
+            Rctx = #rctx{
+                pid_ref = PidRef,
+                type = {coordinator, Verb, [$/ | Path]},
+                nonce = Nonce
+            },
+            track(Rctx),
+            erlang:put(?DELTA_TZ, Rctx),
+            create_resource(Rctx),
+            Rctx
+    end.
+
+set_context_dbname(DbName) ->
+    case is_enabled() of
+        false ->
+            ok;
+        true ->
+            catch case ets:update_element(?MODULE, get_pid_ref(), [{#rctx.dbname, DbName}]) of
```

Comment:
These blocks look very similar in all of the functions. Maybe worth extracting out as a function `update_element(Idx, Value)`. Although in this case you would need to remove one line from stacktrace.